uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-                    next = next  # Prevents incrementing if the current badge already has the highest badge number in the range.
+            if c.NUMBERED_BADGES:
+                if attendee:
+                    return (badge_type in c.PREASSIGNED_BADGE_TYPES or attendee.checked_in)\
+                           and attendee.paid != c.NOT_PAID and not attendee.is_unassigned and attendee.badge_status != c.INVALID_STATUS
                 else:
-                    next += 1
-            else:
-                next = c.BADGE_RANGES[badge_type][0]
+                    return badge_type in c.PREASSIGNED_BADGE_TYPES
+
+        def get_next_badge_num(self, badge_type):
+            """
+            Returns the next badge available for a given badge type. This is essentially a wrapper for auto_badge_num
+            that accounts for new or changed objects in the session.
 
+            :param badge_type: Used to pass to auto_badge_num and to ignore objects in the session that aren't within
+            the badge type's range.
+            :return:
+            """
+            new_badge_num = self.auto_badge_num(badge_type)
             # Adjusts the badge number based on badges in the session
             for attendee in [m for m in chain(self.new, self.dirty) if isinstance(m, Attendee)]:
-                if attendee.badge_type == badge_type and attendee.badge_num <= c.BADGE_RANGES[badge_type][1]:
-                    next = max(next, 1 + attendee.badge_num)
-
-            return next
+                if attendee.badge_type == badge_type and attendee.badge_num is not None\
+                        and attendee.badge_num <= c.BADGE_RANGES[badge_type][1]\
+                        and attendee.badge_num <= self.auto_badge_num(badge_type):
+                    new_badge_num = max(self.auto_badge_num(badge_type), 1 + attendee.badge_num)
 
-        def shift_badges(self, badge_type, badge_num, *, until=None, **direction):
-            # assert_badge_locked()
-            until = until or c.MAX_BADGE
-            assert c.SHIFT_CUSTOM_BADGES
-            assert not any(param for param in direction if param not in ['up', 'down']), 'unknown parameters'
-            assert len(direction) < 2, 'you cannot specify both up and down parameters'
-            down = (not direction['up']) if 'up' in direction else direction.get('down', True)
-            shift = -1 if down else 1
-            for a in self.query(Attendee).filter(Attendee.badge_type == badge_type,
-                                                 Attendee.badge_num >= badge_num,
-                                                 Attendee.badge_num <= until,
-                                                 Attendee.badge_num != 0):
-                a.badge_num += shift
+            return new_badge_num
 
-        def change_badge(self, attendee, badge_type, badge_num=None):
+        def update_badge(self, attendee, old_badge_type, old_badge_num):
```

Comment:
Nope, particularly not in the unit tests. It would often return None, or even worse, the _current_ value. This could very well be a unit test only problem, and I'd definitely be happy to use `orig_value_of` is possible and force the unit tests to work, but I wasn't sure if SQL Alchemy sessions were that reliable.